### PR TITLE
Arad Master Meter water utility meter decoder

### PR DESCRIPTION
**Pull request by avicarmeli**
This decoder decodes the Arad/Master Meter Dialog3G water utility meter.
Decoder returns meter's Serial number and water reading.
Large portion of the payload still to be decoded.

Test results:
```
~/rtl_433/rtl_433_tests$ make test
python3 bin/run_test.py -I time --first-line
WARNING: Ignoring 'tests/inkbird/IBS-P01R/g026_433.92M_250k.cu8'
WARNING: Ignoring 'tests/inkbird/IBS-P01R/g071_433.92M_250k.cu8'
WARNING: Ignoring 'tests/inkbird/IBS-P01R/g001_433.92M_250k.cu8'
WARNING: Ignoring 'tests/inkbird/inkbird_ith20r/g002_433.92M_250k.cu8'
WARNING: Ignoring 'tests/inkbird/inkbird_ith20r/g001_433.92M_250k.cu8'
WARNING: Ignoring 'tests/inkbird/inkbird_ith20r/g003_433.92M_250k.cu8'
WARNING: Ignoring 'tests/scmplus/01/g005_912.6M_2359.3k.cu8'
WARNING: Ignoring 'tests/scmplus/01/g002_912.6M_2359.3k.cu8'
WARNING: Ignoring 'tests/scmplus/01/g003_912.6M_2359.3k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g004_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g009_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g001_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g010_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g007_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g008_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g003_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g005_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g002_867.95M_250k.cu8'
WARNING: Ignoring 'tests/IKEA_Sparsnas/01/g006_867.95M_250k.cu8'
WARNING: Ignoring 'tests/wec_1502/01/gfile002.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g010_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g003_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g005_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g006_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g009_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g008_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g007_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g002_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g001_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/2Gig-DW10/g004_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/RE208/g013_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/RE208/g015_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/RE208/g012_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/RE208/g014_344.975M_250k.cu8'
WARNING: Ignoring 'tests/honeywell/RE208/g011_344.975M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g011_433.92M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g007_433.92M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g005_433.92M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g009_433.92M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g014_433.92M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g013_433.92M_250k.cu8'
WARNING: Ignoring 'tests/adlm_fprf/01/g003_433.92M_250k.cu8'
WARNING: Missing 'tests/idm/IDM/g001_912.6M_2359.3k.cu8'
WARNING: Ignoring 'tests/idm/NetIDM/g012_912.6M_2359.3k.cu8'
WARNING: Ignoring 'tests/idm/NetIDM/g011_912.6M_2359.3k.cu8'
WARNING: Ignoring 'tests/idm/NetIDM/g010_912.6M_2359.3k.cu8'
WARNING: FS20 generated 4 false positive(s) in other decoders: {'Thermopro-TP12'}
WARNING: Proove-Security generated 67 false positive(s) in other decoders: {'Nexa-Security', 'KlikAanKlikUit-Switch'}
WARNING: Nexa-Security generated 125 false positive(s) in other decoders: {'Proove-Security', 'KlikAanKlikUit-Switch'}
WARNING: Abarth-124Spider generated 1 false positive(s) in other decoders: {'Renault'}
WARNING: LaCrosse-R3 generated 4 false positive(s) in other decoders: {'LaCrosse-BreezePro'}
WARNING: LaCrosse-R1 generated 2 false positive(s) in other decoders: {'LaCrosse-TH3'}
WARNING: KlikAanKlikUit-Switch generated 154 false positive(s) in other decoders: {'Proove-Security', 'Nexa-Security'}
WARNING: Acurite-609TXC generated 6 false positive(s) in other decoders: {'Oregon-SL109H'}
WARNING: LaCrosse-TX35DTHIT generated 1 false positive(s) in other decoders: {'LaCrosse-TX29IT'}
WARNING: NETIDM generated 1 false positive(s) in other decoders: {'IDM'}
WARNING: Fineoffset-WH0530 generated 1 false positive(s) in other decoders: {'GT-WT02'}
WARNING: Rubicson-Temperature generated 5 false positive(s) in other decoders: {'Solight-TE44'}
1822 records tested, 0 have failed
```
Running the decoder yields for example:
```
rtl_433 -f 916.2M -R260
rtl_433 version -128-NOTFOUND branch master at 202406241751 inputs file rtl_tcp RTL-SDR

New defaults active, use "-Y classic -s 250k" if you need the old defaults

Detached kernel driver
Found Rafael Micro R820T tuner
[SDR] Using device 0: Realtek, RTL2838UHIDIR, SN: 00000001, "Generic RTL2832U OEM"
Exact sample rate is: 1000000.026491 Hz
[R82XX] PLL not locked!
Allocating 15 zero-copy buffers
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2024-06-26 11:25:37
model     : ARAD-MS-Meter Serial No : 14589391R
Water Reading: 65225.7 M^3                         Integrity : CHECKSUM
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2024-06-26 11:25:44
model     : ARAD-MS-Meter Serial No : 09225790S
Water Reading: 791.5 M^3 Integrity : CHECKSUM
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2024-06-26 11:25:47
model     : ARAD-MS-Meter Serial No : 14895702�
Water Reading: 181.4 M^3 Integrity : CHECKSUM
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2024-06-26 11:25:50
model     : ARAD-MS-Meter Serial No : 09444602S
Water Reading: 757.7 M^3 Integrity : CHECKSUM
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
time      : 2024-06-26 11:25:54
model     : ARAD-MS-Meter Serial No : 00633398�
Water Reading: 232.3 M^3 Integrity : CHECKSUM

```
I am not sure why the decoder is disabled by default and should I change it?

### CODE REVIEW

File: src/devices/arad_ms_meter.c [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/** @file
+    Arad/Master Meter Dialog3G water utility meter.
+    Copyright (C) 2022 avicarmeli
```
Comment (zuckschwerdt) [inline src/devices/arad_ms_meter.c:3]: needs an empty line above and below as in other decoders

File: src/devices/arad_ms_meter.c [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     
+
+    bitbuffer_invert(bitbuffer);
+    
+    //bitbuffer_print(bitbuffer);
+    
+    bitbuffer_extract_bytes(bitbuffer, row,start_pos, mdata, 120);
+  
+    
+    uint32_t serno= mdata[0]| (mdata[1] << 8) | (mdata[2] << 16) | (0 << 24); //24 bit little endian Meter Serial number
```
Comment (zuckschwerdt) [inline src/devices/arad_ms_meter.c:79]: have spaces around `=`, also in the line below

File: src/devices/arad_ms_meter.c [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+};
+
+
+r_device arad_ms_meter = {
+        .name        = "Arad/Master Meter Dialog3G water utility meter",
+        .modulation  = FSK_PULSE_MANCHESTER_ZEROBIT,
+        .short_width = 8.4,
+        .long_width  = 0, //not used
+        .reset_limit = 30,
+        .decode_fn   = &arad_mm_dialog3g_callback,
```
Comment (zuckschwerdt) [inline src/devices/arad_ms_meter.c:119]: should be called `…_decode`

**Comment by zuckschwerdt** [comment]
> I am not sure why the decoder is disabled by default and should I change it?

Only when the checksum is found and working well :) I'll take a peek at the posted codes.

**Comment by zuckschwerdt** [comment]
The presumed checksum is only one byte as seen in [this BitBench](https://triq.net/bitbench#c=%7B177%7D000000003e690aec7ac84b80f98b7300030d0000056f9810d018f8&c=%7B177%7D000000003e690aec7ac84b80f98b7300030d0000056f9810d018f8&c=%7B178%7D000000003e690aec7ac84b80f98b7300030d0000056f9810d018fc&c=%7B178%7D000000003e690aec7ac84b13fe8b73001b050000052880d318f8f8&c=%7B177%7D000000003e690aec7ac84be20a8c7300a41e00000572730b7d08f8&c=%7B176%7D000000003e690aec7ac84be20a8c7300a41e00000572730b7d08f&c=%7B177%7D000000003e690aec7ac84b27108c73006d060000052d17d1da88f8&c=%7B178%7D000000003e690aec7ac84b27108c73006d060000052d17d1da88fc&c=%7B177%7D000000003e690aec7ac84b27108c73006d060000052d17d1da88f8&c=%7B177%7D000000003e690aec7ac84b3ec68c73009d0e000005cc7333ecf0f8&c=%7B177%7D000000003e690aec7ac84b3ec68c73009d0e000005cc7333ecf0f8&c=%7B177%7D000000003e690aec7ac84b63868d7300de0b0000052ae1686bb0f8&c=%7B177%7D000000003e690aec7ac84b63868d7300de0b0000052ae1686bb0f8&c=%7B178%7D000000003e690aec7ac84b45b48d7300ee110000054b1eef6b68fc&c=%7B178%7D000000003e690aec7ac84b45b48d7300ee110000054b1eef6b68fc&c=%7B178%7D000000003e690aec7ac84b6ec98d7300210a00000514db854b40fc&c=%7B178%7D000000003e690aec7ac84b6ec98d7300210a00000514db854b40fc&c=%7B177%7D000000003e690aec7ac84b9da68e0000a2010000058f1243d568f8&c=%7B177%7D000000003e690aec7ac84b9da68e0000a2010000058f1243d568f8&c=%7B176%7D000000003e690aec7ac84b9da68e0000a2010000058f1243d568f&c=%7B177%7D000000003e690aec7ac84bfa1c907300e806000005ea10fef8b8f8&c=%7B178%7D000000003e690aec7ac84bdff7930000700b000005a6bb965e08fc&c=%7B178%7D000000003e690aec7ac84bdff7930000700b000005a6bb965e08fc&c=%7B178%7D000000003e690aec7ac84bdff7930000700b000005a6bb965e08fc&c=%7B178%7D000000003e690aec7ac84bdff7930000700b000005a6bb965e08fc&c=%7B177%7D000000003e690aec7ac84bcf9dde7340acf803000c10d65c6f48f8&c=%7B177%7D000000003e690aec7ac84bcf9dde7340acf803000c10d65c6f48f8&c=%7B177%7D000000003e690aec7ac84bcf9dde7340acf803000c10d65c6f48f8&c=%7B176%7D000000003e690aec7ac84bcf9dde7340acf803000c10d65c6f48f&c=%7B177%7D000000003e690aec7ac86b68e3de734036430500085d997ed2e0f8&c=%7B177%7D000000003e690aec7ac86b68e3de734036430500085d997ed2e0f8&f=P%3A32h%20F%3A56h%20S%3A24d8c%20%3Fhh%20C%3A24d%20%3F12h%20F%3Ah%20CHK%3A8h%20%3Fh&z=1&cw=4).
I.e. `00000000FFFFFFFFFFFFFFSSSSSSSSXXCCCCCCXXXF?????????XFF` is actually `00000000FFFFFFFFFFFFFFSSSSSSSSXXCCCCCCXXXFKKXF` -- can you verify that?

File: src/devices/arad_ms_meter.c (hunk 1 above)
Comment (avicarmeli) [reply src/devices/arad_ms_meter.c:3]: got it

File: src/devices/arad_ms_meter.c (hunk 2 above)
Comment (avicarmeli) [reply src/devices/arad_ms_meter.c:79]: got it

File: src/devices/arad_ms_meter.c (hunk 3 above)
Comment (avicarmeli) [reply src/devices/arad_ms_meter.c:119]: Should it be : &arad_mm_dialog3g_callback_decode?

File: src/devices/arad_ms_meter.c [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    
+    sprintf(sernoout, "%08u%c", serno, mdata[3]-32);
+    
+    float wread = wreadraw;
+    
+    wread=wread/10;
+  
+     /* clang-format off */
+        data = data_make(
+                "model",            "",               DATA_STRING,    "ARAD-MS-Meter",
```
[the remark this answers is not in the document]
Comment (avicarmeli) [reply src/devices/arad_ms_meter.c:92]: got it

File: src/devices/arad_ms_meter.c (hunk 3 above)
Comment (zuckschwerdt) [reply src/devices/arad_ms_meter.c:119]: yes `arad_mm_dialog3g_decode`

**Comment by zuckschwerdt** [comment]
Is it possible to record codes from the same serial over a longer time? Ideally we would get codes with only the count field slowly increasing. Then we could infer what kind of checksum is used.

**Comment by avicarmeli** [comment]
Since the resolution of the water meter is only 100 litter it changes very slow.
At my house we consume about 600 litter per day.
here are some meter flipping readings:

000000003e690aec7ac84bdab18c73006a0e000005aac3d942d8fc
000000003e690aec7ac84bdab18c73006b0e000005cb7b46f478f8

000000003e690aec7ac84bfa1c907300ba060000059e9a639380fc
000000003e690aec7ac84bfa1c907300bb06000005ff22fc2520f8

000000003e690aec7ac84b3320940000de0e00000518adbb8048fe
000000003e690aec7ac84b3320940000df0e00000579152436e8f8


I will try to record some more.

**Comment by avicarmeli** [comment]
It is my first pull request, so please tell me if I need to de anything other then I did.

**Comment by zuckschwerdt** [comment]
Thanks! Basically looks good. I don't think we are going to find the checksum/digest/MAC anytime soon, so this can go in as-is.

There are some code style issues to fix. Use 4 spaces to indent (e.g. clang-format / data_make doesn't look right). Also remove all trailing spaces (as the checker mentions).

The addition of "Credits to all users of mentioned forum below!" is unusual but ok. It should be stand-alone though. E.g. "Additional credits to participants of THE-THREAD"

File: src/devices/arad_ms_meter.c [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+static char *output_fields[] = {
+        "model",
+        "id",
+        "waterread",
+        "mic",
+        NULL,
+};
+
+r_device arad_ms_meter = {
```
Comment (zuckschwerdt) [inline src/devices/arad_ms_meter.c:99]: `r_device const arad_ms_meter = {`

File: src/devices/arad_ms_meter.c [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        "waterread",   "Water Reading",  DATA_FORMAT,    "%.1f M^3", DATA_DOUBLE, wread,
+        //"mic",         "Integrity",      DATA_STRING,    "CHECKSUM",
+        NULL);
+    /* clang-format on */
+
+    decoder_output_data(decoder, data);
+    return 1;
+}
+
+static char *output_fields[] = {
```
Comment (zuckschwerdt) [inline src/devices/arad_ms_meter.c:91]: `static char const *const output_fields[] = {`

**Comment by avicarmeli** [comment]
Sorry for not reading thoroughly the contribute.md.

File: src/devices/arad_ms_meter.c [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    float wread = wreadraw;
+
+    wread = wread / 10;
+
+    /* clang-format off */
+    data = data_make(
+        "model",       "",               DATA_STRING,    "AradMsMeter-Dialog3G",
+        "id",          "Serial No",      DATA_STRING,    sernoout,
+        "waterread",   "Water Reading",  DATA_FORMAT,    "%.1f M^3", DATA_DOUBLE, wread,
```
Comment (zuckschwerdt) [inline src/devices/arad_ms_meter.c:82]: To align with other decoders this should be
`            "volume_m3",    "Volume",       DATA_FORMAT, "%.1f m3", DATA_DOUBLE, wread,`

**Pull request merged**